File: python/tests/test_LocalBrowserLauncher.py
# MODCDP_TRANSLATE_TEST: KEEP THIS TEST FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
# All test cases, descriptions, covered edge cases, and setup should be kept perfectly 1:1 in sync between:
# - ./js/test/test.LocalBrowserLauncher.ts
# - ./go/modcdp/launcher/LocalBrowserLauncher_test.go
# NO MOCKING, NO MONKEY PATCHING, NO SIMULATING, NO FAKING, NO SKIPPING ALLOWED.
# USE REAL USER-FACING CODE PATHS WITH REAL BROWSERS, REAL CLASSES, REAL URLS, etc. Hard fail if keys or other env requirements are missing.
from __future__ import annotations

from collections.abc import Mapping
import tempfile
import unittest
from pathlib import Path

from modcdp.launcher.LocalBrowserLauncher import LocalBrowserLauncher
from modcdp.transport.WSUpstreamTransport import WSUpstreamTransport


class LocalBrowserLauncherTests(unittest.TestCase):
    def test_class_helpers_match_the_local_launcher_surface(self) -> None:
        self.assertIsInstance(LocalBrowserLauncher.findChromeBinary(), str)
        self.assertIsInstance(LocalBrowserLauncher.freePort(), int)

    def test_launches_a_real_browser_over_a_chosen_cdp_port_and_explicit_profile_dir(self) -> None:
        with tempfile.TemporaryDirectory(prefix="modcdp-python-local-profile-") as user_data_dir:
            port = LocalBrowserLauncher.freePort()
            chrome = LocalBrowserLauncher(
                {
                    "launcher_local_headless": True,
                    "launcher_local_chrome_ready_timeout_ms": 45_000,
                    "launcher_local_chrome_ready_poll_interval_ms": 50,
                }
            ).launch({"launcher_local_cdp_listen_port": port, "launcher_local_user_data_dir": user_data_dir})
            cdp_url = chrome.cdp_url
            if cdp_url is None:
                raise AssertionError("expected launcher to return cdp_url")
            transport = WSUpstreamTransport({"upstream_ws_cdp_url": cdp_url})
            transport.connect()

            try:
                self.assertEqual(chrome.cdp_listen_port, port)
                self.assertRegex(cdp_url, rf"^ws://127\.0\.0\.1:{port}/")
                self.assertEqual(chrome.profile_dir, user_data_dir)
                expect_cdp_browser_surface(transport)
            finally:
                transport.close()
                chrome.close()

            self.assertTrue(Path(user_data_dir).exists())

    def test_removes_an_explicit_user_data_dir_when_cleanup_user_data_dir_is_set(self) -> None:
        user_data_dir = tempfile.mkdtemp(prefix="modcdp-python-local-profile-")
        chrome = LocalBrowserLauncher(
            {
                "launcher_local_headless": True,
                "launcher_local_chrome_ready_timeout_ms": 45_000,
            }
        ).launch({"launcher_local_user_data_dir": user_data_dir, "launcher_local_cleanup_user_data_dir": True})

        try:
            self.assertEqual(chrome.profile_dir, user_data_dir)
        finally:
            chrome.close()
        self.assertFalse(Path(user_data_dir).exists())


def expect_cdp_browser_surface(transport: WSUpstreamTransport) -> None:
    version = transport.send("Browser.getVersion")
    expect_version_result(version)

    created = transport.send("Target.createTarget", {"url": "about:blank#modcdp-launcher-test"})
    target_id = created.get("targetId")
    if not isinstance(target_id, str):
        raise AssertionError(f"Target.createTarget result = {created!r}")

    try:
        attached = transport.send("Target.attachToTarget", {"targetId": target_id, "flatten": True})
        session_id = attached.get("sessionId")
        if not isinstance(session_id, str):
            raise AssertionError(f"Target.attachToTarget result = {attached!r}")
        transport.send("Runtime.enable", {}, session_id)
        evaluated = transport.send(
            "Runtime.evaluate",
            {"expression": "(() => ({ ok: true, value: 42 }))()", "returnByValue": True},
            session_id,
        )
        result = object_dict(evaluated.get("result"))
        if result.get("type") != "object" or result.get("value") != {"ok": True, "value": 42}:
            raise AssertionError(f"Runtime.evaluate result = {evaluated!r}")
    finally:
        try:
            transport.send("Target.closeTarget", {"targetId": target_id})
        except Exception:
            pass


def expect_version_result(version: Mapping[str, object]) -> None:
    product = version.get("product")
    if not isinstance(product, str) or ("Chrome" not in product and "Chromium" not in product):
        raise AssertionError(f"Browser.getVersion product = {product!r}")
    if not isinstance(version.get("protocolVersion"), str):
        raise AssertionError(f"Browser.getVersion protocolVersion = {version.get('protocolVersion')!r}")


def object_dict(value: object) -> dict[str, object]:
    if not isinstance(value, Mapping):
        raise AssertionError(f"expected object mapping, got {value!r}")
    return {str(key): raw_value for key, raw_value in value.items()}


if __name__ == "__main__":
    unittest.main()

File: js/test/test.LocalBrowserLauncher.ts
// MODCDP_TRANSLATE_TEST: KEEP THIS TEST FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// All test cases, descriptions, covered edge cases, and setup should be kept perfectly 1:1 in sync between:
// - ./python/tests/test_LocalBrowserLauncher.py
// - ./go/modcdp/launcher/LocalBrowserLauncher_test.go
// NO MOCKING, NO MONKEY PATCHING, NO SIMULATING, NO FAKING, NO SKIPPING ALLOWED.
// USE REAL USER-FACING CODE PATHS WITH REAL BROWSERS, REAL CLASSES, REAL URLS, etc. Hard fail if keys or other env requirements are missing.
import { mkdtemp, rm, stat } from "node:fs/promises";
import { tmpdir } from "node:os";
import path from "node:path";
import { describe, expect, it } from "vitest";

import { LocalBrowserLauncher } from "../src/launcher/LocalBrowserLauncher.js";
import { WSUpstreamTransport } from "../src/transport/WSUpstreamTransport.js";

const LIVE_BROWSER_TIMEOUT_MS = 60_000;

describe("LocalBrowserLauncher", () => {
  it("class helpers match the local launcher surface", async () => {
    expect(LocalBrowserLauncher.findChromeBinary()).toEqual(expect.any(String));
    expect(await LocalBrowserLauncher.freePort()).toEqual(expect.any(Number));
  });

  it(
    "launches a real browser over a chosen CDP port and explicit profile dir",
    { timeout: LIVE_BROWSER_TIMEOUT_MS },
    async () => {
      const userDataDir = await mkdtemp(path.join(tmpdir(), "modcdp-local-profile-"));
      const port = await LocalBrowserLauncher.freePort();
      const chrome = await new LocalBrowserLauncher({
        launcher_local_headless: true,
        launcher_local_chrome_ready_timeout_ms: 45_000,
        launcher_local_chrome_ready_poll_interval_ms: 50,
      }).launch({
        launcher_local_cdp_listen_port: port,
        launcher_local_user_data_dir: userDataDir,
      });
      const cdp = new WSUpstreamTransport({ upstream_ws_cdp_url: chrome.cdp_url });

      try {
        expect(chrome.cdp_listen_port).toBe(port);
        expect(chrome.cdp_url).toEqual(expect.stringMatching(new RegExp(`^ws://127\\.0\\.0\\.1:${port}/`)));
        expect(chrome.profile_dir).toBe(userDataDir);
        await expect(stat(userDataDir)).resolves.toBeTruthy();
        await cdp.connect();
        await expectCdpBrowserSurface(cdp);
      } finally {
        await cdp.close();
        await chrome.close();
        await chrome.close();
        await expectHttpEndpointDown(`http://127.0.0.1:${port}`);
        await expect(stat(userDataDir)).resolves.toBeTruthy();
        await rm(userDataDir, { recursive: true, force: true });
      }
    },
  );

  it(
    "removes an explicit user data dir when cleanup_user_data_dir is set",
    { timeout: LIVE_BROWSER_TIMEOUT_MS },
    async () => {
      const userDataDir = await mkdtemp(path.join(tmpdir(), "modcdp-local-profile-"));
      const chrome = await new LocalBrowserLauncher({
        launcher_local_headless: true,
        launcher_local_chrome_ready_timeout_ms: 45_000,
      }).launch({
        launcher_local_user_data_dir: userDataDir,
        launcher_local_cleanup_user_data_dir: true,
      });

      try {
        expect(chrome.profile_dir).toBe(userDataDir);
        await expect(stat(userDataDir)).resolves.toBeTruthy();
      } finally {
        await chrome.close();
      }

      await expect(stat(userDataDir)).rejects.toMatchObject({ code: "ENOENT" });
    },
  );
});

// MODCDP_TEST_SUPPORT: LANGUAGE-SPECIFIC TEST SUPPORT ONLY.
// Keep the setup semantics above 1:1 with translated tests; helpers here only use real ModCDP transports against real browser endpoints.
async function expectCdpBrowserSurface(cdp: WSUpstreamTransport) {
  const version = await cdp.send("Browser.getVersion");
  expect(version.product).toEqual(expect.stringMatching(/Chrome|Chromium/));
  expect(version.protocolVersion).toEqual(expect.any(String));

  const created = await cdp.send("Target.createTarget", { url: "about:blank#modcdp-launcher-test" });
  expect(created.targetId).toEqual(expect.any(String));
  const targetId = created.targetId as string;

  try {
    const attached = await cdp.send("Target.attachToTarget", { targetId, flatten: true });
    expect(attached.sessionId).toEqual(expect.any(String));
    const sessionId = attached.sessionId as string;
    await cdp.send("Runtime.enable", {}, sessionId);
    const evaluated = await cdp.send(
      "Runtime.evaluate",
      { expression: "(() => ({ ok: true, value: 42 }))()", returnByValue: true },
      sessionId,
    );
    expect(evaluated.result).toMatchObject({ type: "object", value: { ok: true, value: 42 } });
  } finally {
    await cdp.send("Target.closeTarget", { targetId }).catch(() => ({}));
  }
}

async function expectHttpEndpointDown(url: string) {
  await expect
    .poll(
      async () => {
        try {
          await fetch(`${url}/json/version`);
          return false;
        } catch {
          return true;
        }
      },
      { timeout: 5_000, interval: 100 },
    )
    .toBe(true);
}
